### ISSUE TITLE
UB in thread_data::operator delete

### ISSUE THREAD

**Issue by AntonBikineev** [STEllAR-GROUP/hpx#1763]
These lines seem to be wrong as operator delete gets called after thread_data destruction:
```
    void thread_data::operator delete(void *p, std::size_t size)
    {
        HPX_ASSERT(sizeof(thread_data) == size);

        if (0 != p)
        {
            thread_data* pt = static_cast<thread_data*>(p);
            thread_pool* pool = pt->pool_;
            HPX_ASSERT(pool);

            pool->deallocate(pt);
        }
```
https://github.com/STEllAR-GROUP/hpx/blob/master/src/runtime/threads/thread_data.cpp#L67-L75

Hope to have time and fix it soon unless somebody wants to beat me to.

**Comment by hkaiser** [comment]
@AntonBikineev if you know how to fix this, please go ahead.

**Comment by sithhell** [comment]
Maybe call this in `intrusive_ptr_release` instead?

**Comment by AntonBikineev** [comment]
@sithhell Might be as a solution but will tie the use of thread_data with intrusive_pointers only

**Comment by sithhell** [comment]
Am 25.09.2015 4:52 nachm. schrieb "Anton Bikineev" <notifications@github.com
>:
>
> @sithhell Might be as a solution but will tie the use of thread_data with
intrusive_pointers only

As far as I'm aware, that's the case right now, it's only used with
intrusive_ptr.

**Comment by AntonBikineev** [comment]
@sithhell It doesn't seem to be architecturally correct to implicitly require type to be used only with intrusive pointers (counter isn't the case)

**Comment by sithhell** [comment]
Am 25.09.2015 5:13 nachm. schrieb "Anton Bikineev" <notifications@github.com
>:
>
> @sithhell It doesn't seem to be architecturally correct to implicitly
require type to be used only with intrusive pointers (counter isn't the
case)

I kind of agree. In that case, I guess, the design flaw is in the
allocation process itself already, I'm thinking that if you factor out the
deleter, and require that to be called instead of plain delete might solve
the problem.

**Comment by sithhell** [comment]
On 09/25/2015 05:13 PM, Anton Bikineev wrote:
> @sithhell <https://github.com/sithhell> It doesn't seem to be
> architecturally correct to implicitly require type to be used only with
> intrusive pointers (counter isn't the case)

I guess the only non UB way to go is something along those lines:
http://melpon.org/wandbox/permlink/MF1CPJfAMj3BCrTe

**Comment by hkaiser** [comment]
> I guess the only non UB way to go is something along those lines:
> http://melpon.org/wandbox/permlink/MF1CPJfAMj3BCrTe

I agree, but only if you extract the reference to the pool before calling the destructor.

**Comment by K-ballo** [comment]
Could we please try to figure out how the code ought to behave before attempting to fix it? The whole construction looks suspicious.

Consider `bool is_created_from(void* pool) const`, that seems to suggest that the _pool_ is part of the `thread_data` model, and a fundamental part of it as one cannot construct a `thread_data` without a pool. In this scenario, it seems a `thread_data` must be created from a pool by design, yet the design does not enforce this and turn optional hints to an allocation scheme useless. When correctly used, this would lead to situations like `new (p) A(p)`. This should instead be written as `p.new_<A>()` or similar, and the custom `new`/`delete` operators should be completely removed.

On the other hand, it's possible that `is_created_from` crept int due to a bogus implementation of custom `new`/`delete` operators. The way these custom operators handle optional hints is by allocating extra memory to put stuff into it. This data belongs to the raw bits only, it cannot be given to the concrete type being constructed, that's orthogonal. They have to do this in a way that appears redundant, replacing 4, 5, or 6 operators due to language reasons. In that case, `is_created_from` and all mention of _pool_ should be removed from `thread_data`, and the operators should be implemented correctly.

So which is it? Is a _pool_ a fundamental piece of the design of a `thread_data`, or is it an optional hint to a specialized allocation scheme?

**Comment by hkaiser** [comment]
`is_created_from` is used to ensure that the `thread_data` is deleted by the same pool as it was allocated from. And yes, `thread_data` is by design supposed to be created from a pool only. 

Currently, the `thread_data` instances are created like `t = new (p) A(p,...)` (which is correct but could be replaced as suggested by @K-ballo, see [here](https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/runtime/threads/policies/thread_queue.hpp#L217)) and are deleted (through an `intrusive_ptr`) by `delete t` (see [here](https://github.com/STEllAR-GROUP/hpx/blob/master/src/runtime/threads/thread_data.cpp#L38)).

**Comment by hkaiser** [comment]
This has been fixed by merging #1775: Restricting thread_data to creating only with intrusive_pointers
